### PR TITLE
Speaker and event details to firestore

### PR DESCRIPTION
**Pull request by quiro91**
## Problem

We are moving more data to Firestore

## Solution

Speakers and event details now come from Firestore

### Test(s) added 

None unfortunately

### Paired with 

Nobody

### CODE REVIEW

File: app/src/main/java/net/squanchy/schedule/ScheduleService.kt [hunk 1]
```diff
@@ -35,68 +25,23 @@ class FirestoreScheduleService(
...
-            it.name,
-            it.bio,
-            it.companyName.optional(),
-            it.companyUrl.optional(),
-            it.personalUrl.optional(),
-            it.photoUrl.optional(),
-            it.twitterUsername.optional()
-        )
-    }
+    private fun <T, U> combineInAPair(): BiFunction<List<T>, U, Pair<List<T>, U>> = BiFunction(::Pair)
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/schedule/ScheduleService.kt:44]: can you inline this now? or does it lose the types?

File: app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+fun FirestoreEvent.toEvent(checksum: Checksum, dateTime: DateTimeZone, favorites: FirebaseFavorites?) = Event(
+    id = id,
+    numericId = checksum.getChecksumOf(id),
+    startTime = LocalDateTime(startTime),
+    endTime = LocalDateTime(endTime),
+    title = title,
+    place = place?.toPlace().optional(),
+    experienceLevel = experienceLevel.optional().flatMap { ExperienceLevel.tryParsingFrom(it) },
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt]: shouldn't this be in line with others in the form of `experienceLevel?.toExperienceLevel().optional()`?

File: app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            timeZoneObservable,
+            Function3(combineIntoEvent)
+        ).subscribeOn(Schedulers.io())
+    }
+
+    private fun FirestoreVenue.extractTimeZone(): DateTimeZone {
+        return DateTimeZone.forID(timezone)
+    }
+
+    private val combineIntoEvent: (FirestoreEvent, FirebaseFavorites, DateTimeZone) -> Event
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt]: can we make this a function and refer to it with `::` instead?

File: app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        get() = { event, favourites, timezone -> event.toEvent(checksum, timezone, favourites) }
+
+    private val combineIntoEvents: (List<FirestoreEvent>, FirebaseFavorites, DateTimeZone) -> List<Event>
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt]: same here

File: app/src/main/java/net/squanchy/service/repository/firebase/FirestoreSpeakerRepository.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    private val checksum: Checksum
+) : SpeakerRepository {
+
+    override fun speakers(): Observable<List<Speaker>> {
+        return dbService.speakers()
+            .map { it.map { it.toSpeaker(checksum) } }
+    }
+
+    override fun speaker(speakerId: String): Observable<Speaker> {
+        return dbService.speakers()
```
Comment (fourlastor) [inline app/src/main/java/net/squanchy/service/repository/firebase/FirestoreSpeakerRepository.kt]: this has to go with a `dbService.speaker(speakerId)`, we have firestore so we can do querying :)

**Review by fourlastor — CHANGES_REQUESTED** [review]
A few comments, the most important being using querying for the speakers, the rest is mostly nitpicking

File: app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import net.squanchy.service.firestore.model.schedule.FirestoreSpeaker
+import net.squanchy.service.firestore.model.schedule.FirestoreTrack
+import net.squanchy.speaker.domain.view.Speaker
+import net.squanchy.support.lang.Checksum
+import net.squanchy.support.lang.optional
+import net.squanchy.venue.domain.view.Venue
+import org.joda.time.DateTimeZone
+import org.joda.time.LocalDateTime
+
+fun FirestorePlace.toPlace(): Place = Place.create(id, name, floor.optional())
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt]: These should be very easy to test! Can we have tests plz?

File: app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import net.squanchy.service.firestore.FirestoreDbService
+import net.squanchy.service.firestore.model.conferenceinfo.FirestoreVenue
+import net.squanchy.service.firestore.model.schedule.FirestoreEvent
+import net.squanchy.service.firestore.toEvent
+import net.squanchy.service.repository.EventRepository
+import net.squanchy.support.lang.Checksum
+import org.joda.time.DateTimeZone
+
+class FirebaseEventRepository(
+    private val dbService: FirebaseDbService,
```
Comment (rock3r) [inline app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt:18]: :x: using the favourites from RTDB is most likely _not_ going to be working. We should leave them empty for now! There's a TODO for them in #367.

**Review by rock3r — CHANGES_REQUESTED** [review]
Great one, except that we don't have favourites for now and we should NOT use the ones from RTDB as they're an entirely different thing

File: app/src/main/java/net/squanchy/schedule/ScheduleService.kt (hunk 1 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/schedule/ScheduleService.kt:44]: Yeah it does lose the type

File: app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt (hunk 2 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt]: Makes sense 👍 
Edit: the issue here is that experienceLevel is a String and I don't really want to expose an extension function over the String type. Shall we have it private?

File: app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt (hunk 6 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt]: Will do a PR on top of this one for testing, working on it

File: app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt (hunk 2 above)
Comment (fourlastor) [reply app/src/main/java/net/squanchy/service/firestore/FirestoreMappers.kt]: Yeah private is cool to me

**Comment by quiro91** [comment]
@rock3r @fourlastor ready for another review. Also, before merging take a look at #453 which adds test for a lot of the Firestore model

File: app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt (hunk 7 above)
Comment (quiro91) [reply app/src/main/java/net/squanchy/service/repository/firebase/FirebaseEventRepository.kt:18]: Fixed

**Comment by rock3r** [comment]
Do not merge until #453 is shipped

**Review by rock3r — APPROVED** [review]
Nice one, just waiting for the tests to be merged and then we can ship for what I am concerned

**Review by fourlastor — APPROVED** [review]

**Comment by rock3r** [comment]
Tests have been merged, now we can 🚢 once green

**Pull request merged**